src/Lumper.Lib/BSP/BspFile.cs

```diff
@@ -35,45 +43,238 @@ public class BspFile
     // creating a BspFile instance without loading an actual BSP.
     private BspFile() { }
 
-    public BspFile(string path) => Load(path);
+    public static BspFile? FromPath(IoHandler handler, string path)
+        => new BspFile { Name = Path.GetFileNameWithoutExtension(path) }
+            .Load(handler, path);
 
-    public void Load(string path)
+    public static BspFile? FromStream(IoHandler handler, Stream stream)
+        => new BspFile()
+            .Load(handler, stream);
+
+    public BspFile? Load(IoHandler handler, string path)
     {
-        // TODO: loads of error handling
         Name = Path.GetFileNameWithoutExtension(path);
-        var filePath = Uri.UnescapeDataString(Path.GetFullPath(path));
-        using FileStream stream = File.OpenRead(filePath);
-        Load(stream);
-        FilePath = filePath; // Set this at the end because Load(stream) resets it
+        FilePath = GetUnescapedFilePathString(path);
+        FileStream = File.OpenRead(FilePath);
+
+        using var reader = new BspFileReader(this, FileStream, handler);
+
+        // UI/CLI should track whether this gets cancelled and dispose of this class, it's in a useless
+        // half-loaded state and mustn't be used.
+        if (!reader.Load())
+        {
+            Logger.Info("Loading cancelled");
+            return null;
+        }
+
+        Logger.Info($"Loaded {Name}.bsp");
+        return this;
     }
 
-    public void Load(Stream stream)
+    public BspFile? Load(IoHandler handler, Stream stream)
     {
-        FilePath = null;
-        Stream.Dispose();
-        Stream = new MemoryStream();
-        stream.CopyTo(Stream);
-        var reader = new BspFileReader(this, Stream);
-        reader.Load();
+        using var reader = new BspFileReader(this, stream, handler);
+
+        if (!reader.Load())
+        {
+            Logger.Info("Loading cancelled");
+            return null;
+        }
+
+        Logger.Info("Loaded BSP from stream");
+        return this;
     }
 
-    public void Save(string path)
+    public void SaveToFile(
+        IoHandler handler,
+        string? path,
+        DesiredCompression compress,
+        bool makeBackup = true)
     {
-        if (path == FilePath)
-            throw new IOException("Can't write BSP to the same file");
-        using FileStream stream = File.Open(path, FileMode.Create);
-        Save(stream);
+        // If you add something here, also add it to the BspReader
+        if (path is null && FilePath is null)
+            throw new ArgumentException("Not given a path to write to, and current BSP doesn't have a path");
+
+        string outPath;
+        string? backupPath = null;
+        string? escapedPath;
+        if (path is not null &&
+            (escapedPath = GetUnescapedFilePathString(path)) != FilePath)
+        {
+            outPath = escapedPath;
+        }
```

Comment:
why compare the escapedPath and FilePath .. if path isn't null we can just escape and use it, right?

Reply:
Because of the below backup code, I'm ruling out the cases where we *don't* need to make a backup first. But readability is crap, I'll restructure it.
```cs
    if (path is not null &&
            (escapedPath = GetUnescapedFilePathString(path)) != FilePath)
        {
            outPath = escapedPath;
        }
        else
        {
            ... backup stuff ...
```